src/compiler/compiler.ts

```diff
@@ -0,0 +1,168 @@
+    return Array.from(this.files.values()).map(f => f.path);
+  }
+  getScriptVersion(path: string) {
+    const f = this.get(path);
+    if (f === undefined) {
+      return '';
+    }
+    return f.version.toString();
+  }
+}
+
+export class Compiler {
+  private options?: ts.CompilerOptions;
+  private service: ts.LanguageService;
+  private files: Files;
+
+  constructor(options?: ts.CompilerOptions) {
+    this.files = new Files();
+    this.options = options;
+    this.service = this.createServiceHost();
+  }
+  setOptions(options: ts.CompilerOptions) {
+    this.options = options;
+  }
+  private createServiceHost() {
+    const that = this;
+
+    let service: ts.LanguageService;
+
+    class ServiceHost implements ts.LanguageServiceHost {
+      // getCustomTransformers() {
+      //     return {
+      //         before: [tsruntimeTransformer(service.getProgram())]
+      //     }
+      // }
+      getScriptFileNames() {
+        return that.files.getFileNames();
+      }
+      getScriptVersion(fileName: string) {
+        return that.files.getScriptVersion(fileName);
+      }
+      getScriptSnapshot(fileName: string) {
+        const file = that.files.get(fileName);
+        if (file !== undefined) {
+          return file.snapshot;
+        }
+        if (!fs.existsSync(fileName)) {
+          return undefined;
+        }
+        return ts.ScriptSnapshot.fromString(
+          fs.readFileSync(fileName).toString(),
+        ); //todo maybe put to files cache
+      }
+      getCurrentDirectory = () => process.cwd();
+      getCompilationSettings = () => that.options;
+      getDefaultLibFileName = (options: ts.CompilerOptions) =>
+        ts.getDefaultLibFilePath(options);
+      fileExists = ts.sys.fileExists;
+      readFile = ts.sys.readFile;
+      readDirectory = ts.sys.readDirectory;
+      // resolveTypeReferenceDirectives(typeDirectiveNames: string[], containingFile: string) {
+      //     const resolved = typeDirectiveNames.map(directive =>
+      //         ts.resolveTypeReferenceDirective(directive, containingFile, options, ts.sys)
+      //             .resolvedTypeReferenceDirective);
+
+      //     // resolved.forEach(res => {
+      //     if (res && res.resolvedFileName) {
+      //         fileDeps.add(containingFile, res.resolvedFileName);
+      //     }
+      // });
+
+      // return resolved;
+      // }
+    }
+
+    service = ts.createLanguageService(
+      new ServiceHost(),
+      ts.createDocumentRegistry(),
+    );
+
+    return service;
+  }
+
+  emitFile({ path, src }: { path: string; src: string }) {
+    this.files.update({ path, src });
+
+    let output = this.service.getEmitOutput(path);
+
+    if (output.emitSkipped) {
+      const errors = this.formatErrors(path);
+      throw new Error(`Errors while compile ts: \n${errors}`);
+    }
+    const res = utils.findResultFor(path, output);
+    return res;
+  }
+
+  private formatErrors(fileName: string) {
+    let allDiagnostics = this.service
```

Comment:
Nice! Though maybe you can test it on a system with loads of files? I was trying to do this as well over the past 2 days, and so far tests where like 10x slower when you have a lot of files and resolutions let's say over 5K files.

I can through download your branch and test this out, if you'd like. Probably tomorrow.

Reply:
I think this is not main problem.
In ts-jest master currently you have smth like
fileSources.map(s => transpileModiule(s)
And transpile module don't catch any imported files. This means that it could be easilly paralellized.
With program api it looks like we can't paralelize build. You could try to run tests with this branch - it works faster in single thread mode, than with multithreading. It recompiling same files in different processes. I don't know how to fix this. Maybe it could be common shared storage of snapshots.